File: tests/unit/test_matrix_arithmetic.js
/* Arithmetic inference is a finite signature lookup, never value evaluation. */
const assert=require('node:assert/strict'),fs=require('node:fs'),vm=require('node:vm'),path=require('node:path');
const payload=JSON.parse(fs.readFileSync(0,'utf8')),dir=path.resolve(__dirname,'../../src/editor');
const stub={addEventListener(){},classList:{toggle(){}},value:''};
const context=vm.createContext({assert,payload,console,crypto:globalThis.crypto,location:{pathname:'/',hash:''},history:{replaceState(){}},window:{addEventListener(){},getSelection(){return null;}},document:{addEventListener(){},querySelector:()=>stub,querySelectorAll:()=>[]},sessionStorage:{getItem(){return '';},setItem(){}},setTimeout(){return 1;},clearTimeout(){}});
for(const name of ['functions_model.js','functions_ui.js','graph_ui.js','inspector.js'])vm.runInContext(fs.readFileSync(path.join(dir,name),'utf8'),context);
const app=fs.readFileSync(path.join(dir,'app.js'),'utf8');vm.runInContext(app.slice(0,app.indexOf("$('#canvas').addEventListener('dragover'")),context);
vm.runInContext(`
catalog=payload.catalog;setTypeContract(payload.contract);editorTarget='top';stage='pixel';graphTrail=[];
const same=(a,b)=>assert.equal(JSON.stringify(a),JSON.stringify(b));
const make=(key,id,params={},auto=false)=>({id,definitionUuid:'sgrape.builtin.'+key,params:{...catalog.find(d=>d.key===key).defaults,...params},ui:auto?{typeMode:'auto'}:{}});
const source=(type,id)=>isMatrixType(type)?make('matrix',id,{type,values:shapedValue(1,type)}):typeComponents(type)>1?make('vector',id,{type,components:[1,2,3,4]}):make('scalar',id,{type,value:2});
const documentFor=(key,a,b,params={},auto=true)=>({declarations:[],functions:[],stages:{pixel:{nodes:[source(a,'a'),source(b,'b'),make(key,'op',params,auto)],edges:[{from:['a','out'],to:['op','a']},{from:['b','out'],to:['op','b']}]}}});
let count=0;
for(const d of catalog.filter(isArithmetic))for(const variant of typeVariants(d)){
  graph=documentFor(d.key,variant.inputs.a,variant.inputs.b);
  const before=clone(graph),plan=planAutoGraph(graph,current());
  assert.equal(plan.issues.size,0);same(plan.ports.get('op'),{inputs:variant.inputs,outputs:variant.outputs});
  resolveAutoEdit(graph,{...clone(graph),stages:{pixel:{nodes:[],edges:[]}}});
  same(concretePorts(graph,current().nodes[2],null),plan.ports.get('op'));
  const saved=JSON.stringify(graph);graph=JSON.parse(saved);same(concretePorts(graph,current().nodes[2],null),plan.ports.get('op'));
  same(before.stages.pixel.nodes[2].params.type,d.defaults.type);count++;
}
for(const key of ['add','subtract','multiply','divide']){
  const n=make(key,'op',{type:'mat2x3'}),type=concretePorts({declarations:[]},n,null).inputs.b;
  same(defaultInput(n,'b',type),key==='multiply'?shapedValue(1,type):Array(typeComponents(type)).fill(key==='divide'?1:0));
}
for(const port of ['a','b']){
  graph=documentFor('multiply','mat2x3','float');current().nodes[port==='a'?0:1]=source('mat2x3',port);
  current().edges=current().edges.filter(e=>e.to[1]===port);
  assert.equal(planAutoGraph(graph,current()).ports.get('op').outputs.out,'mat2x3');
}
graph=documentFor('multiply','mat2x3','vec2',{type:'vec3'},false);
let plan=planAutoGraph(graph,current());same(plan.ports.get('op'),{inputs:{a:'mat2x3',b:'vec2'},outputs:{out:'vec3'}});
graph=documentFor('add','mat3','vec3');assert.throws(()=>planAutoGraph(graph,current()));
graph=documentFor('multiply','mat2x3','vec3');assert.throws(()=>planAutoGraph(graph,current()));
graph=documentFor('add','mat3','float');
resolveAutoEdit(graph,{...clone(graph),stages:{pixel:{nodes:[],edges:[]}}});
const mixed=clone(graph),op=current().nodes[2];op.inputValues={a:shapedValue(1,'mat3'),b:7};
reshapeTypedInputs(op,catalog.find(d=>d.key==='add'),'mat3',{a:'mat3',b:'mat3'});assert.equal(op.inputValues.b.length,9);
reshapeTypedInputs(op,catalog.find(d=>d.key==='add'),'mat3',{a:'mat3',b:'float'});assert.equal(op.inputValues.b,7);
graph=clone(mixed);const previous=clone(graph);current().nodes[1]=source('mat3','b');resolveAutoEdit(graph,previous);
assert.equal(current().nodes[2].params.operandTypes.b,'mat3');
graph=clone(mixed);assert.equal(concretePorts(graph,current().nodes[2],null).inputs.b,'float');
// A downstream array must follow the result vector, not either matrix shape.
graph=documentFor('multiply','mat2x3','vec2');current().nodes.push(make('array_create','array',{elementType:'vec3'}));current().edges.push({from:['op','out'],to:['array','value']});
plan=planAutoGraph(graph,current());assert.equal(plan.ports.get('array').outputs.out,'vec3[4]');
console.log('matrix arithmetic editor passed: '+count+' signatures');
`,context);

File: tests/unit/test_matrix_arithmetic.py
"""Operator contracts, concrete saved signatures, and unchanged legacy shaders."""
import copy
import json
from pathlib import Path
import subprocess
import unittest

import sgrape_core as c
from test_matrix_foundation import graph


class MatrixArithmetic(unittest.TestCase):
    def test_every_signature_compiles_and_round_trips_without_graph_mutation(self):
        for key in c.ARITHMETIC_KEYS:
            self.assertEqual(c.arithmetic_variants(key)[0]['type'],c.CATALOG[key]['defaults']['type'])
            for variant in c.arithmetic_variants(key):
                with self.subTest(key=key,inputs=variant['inputs']):
                    operation=c.node(key,'matrix',type=variant['type'],**variant.get('params',{}))
                    operation['inputValues']={p:c.filled_value(t,2) for p,t in variant['inputs'].items()}
                    document=graph([operation],ty=variant['type']);before=copy.deepcopy(document)
                    result=c.compile_graph(document)
                    self.assertEqual(document,before)
                    self.assertEqual(result,c.compile_graph(json.loads(json.dumps(document))))
                    self.assertEqual(c.resolved_ports(c.CATALOG[key],operation['params']),{k:variant[k] for k in ('inputs','outputs')})

    def test_operand_order_and_shape(self):
        for family,prefix,vector in [('float','mat','vec'),('double','dmat','dvec')]:
            for key in c.ARITHMETIC_KEYS:
                for a,b,out in [(prefix+'3',family,prefix+'3'),(family,prefix+'3',prefix+'3')]:
                    self.assertEqual(c.arithmetic_result(key,a,b),out)
            for a,b,out in [(prefix+'2x3',prefix+'4x2',prefix+'4x3'),(prefix+'2x3',vector+'2',vector+'3'),(vector+'3',prefix+'2x3',vector+'2')]:
                self.assertEqual(c.arithmetic_result('multiply',a,b),out)
            self.assertIsNone(c.arithmetic_result('multiply',prefix+'2x3',prefix+'2x3'))
            for key in ('add','subtract','divide'):
                self.assertIsNone(c.arithmetic_result(key,prefix+'3',vector+'3'))
                self.assertIsNone(c.arithmetic_result(key,prefix+'2x3',prefix+'3x2'))

    def test_invalid_saved_signatures_are_rejected(self):
        for operands in [None,[],{'a':'mat3'},{'a':'mat3','b':'vec3'}, {'a':'mat3','b':'dmat3'}, {'a':'mat3','b':{}}, {'a':'mat3','b':'float','c':'float'}]:
            with self.subTest(operands=operands),self.assertRaises(c.GraphError):
                c.resolved_ports(c.CATALOG['add'],{'type':'mat3','operandTypes':operands})
        with self.assertRaises(c.GraphError):c.resolved_ports(c.CATALOG['multiply'],{'type':'mat3','operandTypes':{'a':'mat2x3','b':'mat4x2'}})

    def test_scalar_is_not_emitted_as_diagonal_matrix(self):
        for key in c.ARITHMETIC_KEYS:
            for a,b in [('mat3','float'),('float','mat3')]:
                operation=c.node(key,'matrix',type='mat3',operandTypes={'a':a,'b':b})
                operation['inputValues']={'a':c.filled_value(a,2),'b':c.filled_value(b,3)}
                code=c.compile_graph(graph([operation],ty='mat3'))['pixel']
                scalar='2.0' if a=='float' else '3.0'
                self.assertNotIn('mat3('+scalar+')',code)
                self.assertIn(('('+scalar+' ' if a=='float' else ' '+scalar+')'),code)

    def test_unconnected_matrix_defaults_match_operator_neutral_values(self):
        for ty in c.MATRIX_TYPES:
            for key in ('add','subtract'):self.assertEqual(c.input_default(key,'b',ty),[0]*c.type_components(ty))
            self.assertEqual(c.input_default('divide','b',ty),[1]*c.type_components(ty))
            self.assertEqual(c.input_default('multiply','b',ty),c.matrix_identity(ty))

    def test_editor_signatures_wiring_history_and_local_creator(self):
        payload={'catalog':list(c.CATALOG.values()),'contract':c.type_contract()}
        output=subprocess.check_output(['node',str(Path(__file__).with_name('test_matrix_arithmetic.js'))],input=json.dumps(payload),text=True)
        self.assertIn('matrix arithmetic editor passed',output)
